### ISSUE TITLE
Cannot use different types for the fields (with the same name) in multiple templates

### ISSUE THREAD

**Issue by jamespohalloran** [tinacms/tinacms#3150]
## Description

When you have a field that has the same field name of another field in another template, the build will fail if the fieldType doesn't match exactly.

Originally logged [here](https://github.com/tinacms/tinacms/discussions/2249#discussioncomment-1710982) by @Perni1984 . 

### Steps to reproduce

Create the following field:

```
        {
          type: "object",
          label: "Foo",
          name: "foo",
          templates: [
            {
              label: "Blog Posts",
              name: "post1",
              fields: [
                {
                  type: "string",
                  label: "title",
                  name: "title",
                  required: true,
                },
              ],
            },
            {
              label: "Blog Posts",
              name: "post2",
              fields: [
                {
                  type: "string",
                  label: "title",
                  name: "title",
                },
              ],
            },
          ],
        },
```

^ One is required, one is not

### Expected result

Build should pass (Not exactly sure what the types would look like to allow this)

### Actual result

Build fails with:

> Error 3: GraphQLDocumentError: Fields "title" conflict because they return conflicting types "String!" and "String". Use different aliases on the fields to fetch both if this was intentional.

#### Tinacms Package Versions

├── @tinacms/cli@0.61.9
├── tinacms@0.69.4

**Comment by cereallarceny** [comment]
Subscribed to this.

**Comment by jeffsee55** [comment]
> Why not just include the template name in the created type

The issue isn't the type (which is `String` vs `String!`), it's the field name being the same. To avoid this, we'd need to change the field from `title` to something like `blog1_title`, which would be pretty rough to deal with.

To be clear, there's no issue building the schema with this, it's only an issue with our autogenerated client. We can probably find some clever workarounds if we invest more time in that (ie. using aliases under the hood).

**Comment by cereallarceny** [comment]
@jeffsee55 Do you know of a temporary work around? I'm somewhat blocked here. My initial idea was maybe to wrap all templates in an object that has a unique name, so it would force all generated field names to be unique. However, I'm not familiar enough with the autogenerated client to know that it would make a tangible difference.

**Comment by jeffsee55** [comment]
That would work but I would expect that just using a different field name would be the most straightforward workaround, is that not an option for you?

**Comment by cereallarceny** [comment]
In my case, not without rewriting most of my front-end. It's expecting a certain set of keys. Plus, each of the pages have similar components, but they each have slightly deviating responsibilities.

**Comment by jeffsee55** [comment]
This doesn't sound like something GraphQL will ever change, _a lot_ of people don't like it https://github.com/graphql/graphql-js/issues/53, myself included. We will continue to discuss a built-in solution to this but for now there's nothing we can do.

**Comment by Perni1984** [comment]
@jeffsee55 imho the most straightforward option to solve this would be the ability to support aliases in the tinacms schema. With this people have the ability to solve this immediately when the situation occurs. The Alias should then be picked up by the autogenerated client (see https://discord.com/channels/835168149439643678/835172431836479508/900617826572836904).

Eg. something like

```
        {
          type: "object",
          label: "Foo",
          name: "foo",
          templates: [
            {
              label: "Blog Posts",
              name: "post1",
              alias: "post1"
              fields: [
                {
                  type: "string",
                  label: "title",
                  name: "title",
                  alias: "requiredTitle"
                  required: true,
                },
              ],
            },
            {
              label: "Blog Posts",
              name: "post2",
              fields: [
                {
                  type: "string",
                  label: "title",
                  name: "title",
                },
              ],
            },
          ],
        },
```

For us this is a real blocker aswell and we try to "avoid" templates whenever possible (as you don't know beforehand when you run into such a situation as the schema can change due to client requirements). We're missing out an important feature of tina I guess. As the autogenerated client is now more usable and stable than it was several month ago it is our go to strategy for querying tina and not using any self written queries.

**Comment by cereallarceny** [comment]
Love the solution @Perni1984 is proposing with self-declared aliases. Seems like that would be a way to more or less permanently avoid having to wait on a solution upstream from GQL. This would solve my problem, and a variety of problems I've seen other users reporting related to `name`, `required` and `type` conflicts.

**Comment by jeffsee55** [comment]
@Perni1984 this is definitely something we could do but I'm sort of confused on what value it offers, why wouldn't you just call your field `requiredTitle`?

**Comment by cereallarceny** [comment]
Because the issue doesn't come up only in the presence of required fields @jeffsee55. It also happens when the Scalar value of the field is different. Sometimes title is a string, sometimes it's a rich-text, and it could even be an object that breaks down into multi-line titles. This isn't only a problem with `required`, it's a problem with `type` and any other key in the schema definition that would make the field act differently.

**Comment by cereallarceny** [comment]
Correct. I could rename all my conflicting fields, which I have done. But
in my opinion, I think this should “just work”. To have the implementer
needing to consider key naming conflicts, when they might or might not
trigger, at what depth they are happening, isn’t really ideal. Sure,
renaming the field will work, but I don’t feel like this is the most ideal
solution for those using Tina.

**Comment by Perni1984** [comment]
> @cereallarceny I understand that but I'm still not sure how it's different than the name of the field itself? If you're willing to work with a union of:
> 
> ```js
> [
>   { post1: { requiredTitle: "hello" } },
>   { post2: { title: "hello, again" } }
> ]
> ```
> 
> There's no difference than if the actual field _name_ of `post1`'s title field was just `requiredTitle`?

I was under the assumption that the alias only is used at the GraphQL level, but the autogenerated client would still return then something like 

```js
[
  { post1: { title: "hello" } },
  { post2: { title: "hello, again" } }
]
```

even if underneath for post1 GraphQL queries `requiredTitle` alias would be used.

**Comment by jeffsee55** [comment]
@Perni1984 I understand, so you still need the shape to be such that _both_ templates have a `title` key, that makes sense to me. 

I don't think it's something we could do very easily with our current implementation, cc @jamespohalloran this would be doable in the more advanced client we've talked about.

**Comment by jeffsee55** [comment]
And if we did that under the hood there's probably no need to make the developer provide an alias property, we can do that automatically

**Comment by Perni1984** [comment]
> And if we did that under the hood there's probably no need to make the developer provide an alias property, we can do that automatically

even better ;-)

**Comment by jeffsee55** [comment]
not stale

**Comment by JoshStwrt** [comment]
Not stale

**Comment by lassejaco** [comment]
just +1ing here - super annoying, basically have to refactor my entire setup months into using it because my keys keep clashing

**Comment by Had3r** [comment]
Any updates on this? Maybe someone has a workaround that can be shared.

**Comment by buzzb0x** [comment]
Hi folks, wondering if there's any update about this? I don't mind renaming my fields to avoid collisions but I would rather not. Cheers!
